File: packages/schemas/varuna_schemas/models/route.py
"""Flood-safe routing, reachability and the road-conditions feed
(``POST /v1/route``, ``GET /v1/reachability``, ``GET /v1/feeds/road-conditions``;
CLAUDE.md 7.4, 11.9, 12; blueprint 9.3).
"""

from __future__ import annotations

from typing import Literal

from pydantic import Field, computed_field, model_validator

from varuna_schemas.constants import (
    PROFILE_RISK_TOLERANCE,
    PROFILE_THRESHOLDS_CM,
    REACH_COLLAPSE_RATIO,
    VehicleProfile,
)
from varuna_schemas.models.common import (
    IdStr,
    Latitude,
    LineString,
    Longitude,
    LonLat,
    MultiPolygon,
    Polygon,
    Probability,
    Timestamp,
    VarunaModel,
)
from varuna_schemas.models.hotspot import FacilityKind

RouteConfidence = Literal["high", "medium", "low"]
"""Falls with lead time: high below 60 min, medium to 120 min, low beyond (Sky skill decays)."""

RouteLabel = Literal["VARUNA", "Naive (shortest)", "Alternate"]
RoadStatus = Literal["passable", "degraded", "impassable"]
ReachMinutes = Literal[5, 10, 15]


class RouteRequest(VarunaModel):
    """Body of ``POST /v1/route``."""

    origin: LonLat = Field(description="[lon, lat], e.g. KEM Hospital [72.8419, 19.0035].")
    destination: LonLat = Field(description="[lon, lat], e.g. Sion Hospital [72.8628, 19.0176].")
    depart_at: Timestamp | None = Field(
        default=None, description="Departure (IST); None = the console's scrub time."
    )
    profile: VehicleProfile = "car"
    risk_tolerance: Probability | None = Field(
        default=None, description="Max P(impassable) accepted per segment; None = profile default."
    )
    run_id: str | None = Field(default=None, description="None = the latest published run.")
    alternates: int = Field(default=2, ge=0, le=2, description="Number of alternates to return.")
    origin_name: str | None = None
    destination_name: str | None = None
    spread: bool = Field(
        default=True,
        description="Return up to three safe corridors and an assignment (TECH_SPEC 3.2-3.3).",
    )
    trip_id: str | None = Field(
        default=None,
        description=(
            "The client's own stable id for this trip, so the corridor assignment survives a "
            "reload. Nothing is stored against it; None means the request is not spread."
        ),
    )
    explain: bool = Field(default=True, description="Include structured reasons.")

    @property
    def effective_risk_tolerance(self) -> float:
        return (
            self.risk_tolerance
            if self.risk_tolerance is not None
            else PROFILE_RISK_TOLERANCE[self.profile]
        )

    @property
    def threshold_cm(self) -> int:
        return PROFILE_THRESHOLDS_CM[self.profile]


class AvoidedSegment(VarunaModel):
    """A segment the naive route uses that VARUNA avoids, with why."""

    segment_id: IdStr
    name: str = Field(description="e.g. Hindmata junction")
    reached_ts: Timestamp = Field(description="When the naive route would have reached it (IST).")
    threshold_cm: int = Field(gt=0, description="Profile threshold applied.")
    p_exceed: Probability = Field(description="P(depth > threshold) at reached_ts.")
    depth_p50_cm: float | None = Field(default=None, ge=0)
    closed_reason: str | None = Field(
        default=None,
        description=(
            "Set when an authority closed the street rather than the forecast refusing it; the "
            "officer's own words, never a sentence composed by the API."
        ),
    )


class RouteResult(VarunaModel):
    """One route with its forecast-aware summary."""

    label: RouteLabel = "VARUNA"
    geometry: LineString
    eta_min: float = Field(ge=0, description="Travel time in minutes under the forecast.")
    distance_km: float = Field(ge=0)
    max_expected_depth_cm: float = Field(ge=0, description="Max p50 depth met along the route.")
    safe_until: Timestamp | None = Field(
        default=None,
        description="Latest departure for which the route stays passable; None = horizon.",
    )
    arrival_ts: Timestamp | None = None
    segment_ids: list[IdStr] = Field(default_factory=list)
    max_p_exceed: Probability | None = Field(
        default=None, description="Highest P(impassable) on any segment of the route."
    )


class Corridor(VarunaModel):
    """One of up to three safe roads a request may be spread across (TECH_SPEC 3.3, task D-08).

    The corridors are real - each is a different road and each clears the profile's threshold on
    this run. The ``share`` is **policy**: demand is not measured anywhere in this prototype, and
    every response that carries corridors also carries that disclosure in ``notes``.
    """

    id: str = Field(description="Stable per road: a digest of the corridor's segment sequence.")
    label: Literal["A", "B", "C"]
    route: RouteResult
    share: Probability = Field(description="Fraction of requests policy sends this way.")
    assigned: bool = Field(description="True on the corridor this request was assigned to.")
    capacity_score: float = Field(
        ge=0, description="sum(lanes * (1 - congestion_proxy(depth))) over the corridor's edges."
    )
    max_p_exceed: Probability | None = Field(
        default=None, description="Highest P(impassable) met on this corridor."
    )


ReasonKind = Literal["avoided", "design", "timing", "closure"]
"""The four reasons a route can give for going the way it did (UI_SPEC 4)."""


class RouteReason(VarunaModel):
    """One structured reason. **Never prose** - the frontend words it (TECH_SPEC 3.2).

    Fields not relevant to a ``kind`` are absent. A reason whose number is missing is not
    emitted at all, so a screen never has to soften one into "this road may flood".
    """

    kind: ReasonKind
    segment_id: IdStr
    name: str
    depth_cm: float | None = Field(default=None, ge=0)
    threshold_cm: float | None = Field(default=None, gt=0)
    at: Timestamp | None = Field(default=None, description="The instant the number belongs to.")
    probability: Probability | None = None
    dry_until: Timestamp | None = Field(
        default=None, description="timing: last step under 5 cm on the road actually taken."
    )
    dry_below_cm: float | None = Field(default=None, gt=0)
    design_intensity_mm_h: float | None = Field(
        default=None, gt=0, description="design: what the drain under this street was sized for."
    )
    forecast_peak_mm_h: float | None = Field(
        default=None, ge=0, description="design: this run's peak AOI-mean rain, not local rain."
    )
    reason: str | None = Field(default=None, description="closure: the officer's own words.")
    user: str | None = Field(default=None, description="closure: who entered it.")
    until: Timestamp | None = Field(default=None, description="closure: expiry, if one was set.")


class RouteResponse(VarunaModel):
    """Response of ``POST /v1/route`` (blueprint 9.3 shape, with the naive route for comparison)."""

    run_id: str
    valid_ts: Timestamp = Field(description="Cycle time of the run used (IST).")
    profile: VehicleProfile
    depart_at: Timestamp
    risk_tolerance: Probability
    route: RouteResult
    naive: RouteResult | None = Field(
        default=None, description="Shortest route ignoring the forecast."
    )
    avoided: list[AvoidedSegment] = Field(default_factory=list)
    alternates: list[RouteResult] = Field(default_factory=list)
    corridors: list[Corridor] = Field(
        default_factory=list, description="Up to three safe roads and the share policy gives each."
    )
    reasons: list[RouteReason] = Field(
        default_factory=list, description="Structured, never prose; the frontend words them."
    )
    trip_id: str | None = Field(
        default=None, description="Echoed back so a client can confirm which id was assigned."
    )
    notes: list[str] = Field(
        default_factory=list,
        description=(
            "Honesty labels that must travel with the answer: which probability answered, the "
            "spreading disclosure, and any authority closure applied (CLAUDE.md rule 6)."
        ),
    )
    confidence: RouteConfidence
    confidence_note: str = Field(description="e.g. 'high (lead 30 min)'.")
    explanation: str | None = Field(
        default=None,
        description="Plain language, e.g. 'Avoids Hindmata (82 % above 45 cm at 18:10) via Bharatmata'.",
    )
    compute_ms: int = Field(ge=0, description="Routing wall-clock (budget 300 ms).")

    @computed_field  # type: ignore[prop-decorator]
    @property
    def eta_delta_min(self) -> float | None:
        """Extra minutes versus the naive route (positive = VARUNA is slower but safer)."""
        return None if self.naive is None else self.route.eta_min - self.naive.eta_min


class Isochrone(VarunaModel):
    """Area reachable from a facility within ``minutes`` under the forecast."""

    minutes: ReachMinutes
    polygon: Polygon | MultiPolygon | None = Field(
        default=None, description="Concave hull of reached nodes; None when nothing is reachable."
    )
    area_km2: float = Field(ge=0)
    reachable_nodes: int | None = Field(default=None, ge=0)


class ReachabilityResponse(VarunaModel):
    """``GET /v1/reachability`` and one feature group of ``reachability.geojson``."""

    run_id: str
    valid_ts: Timestamp = Field(description="Cycle time of the run (IST).")
    t: Timestamp = Field(description="Forecast slice the isochrones are computed for (IST).")
    facility_id: IdStr
    facility_name: str = Field(description="e.g. KEM Hospital, Parel")
    facility_kind: FacilityKind
    lon: Longitude
    lat: Latitude
    profile: VehicleProfile
    isochrones: list[Isochrone] = Field(default_factory=list)
    dry_area_15_km2: float = Field(ge=0, description="15-min catchment area with no flooding.")

    @model_validator(mode="after")
    def _one_per_level(self) -> ReachabilityResponse:
        levels = [iso.minutes for iso in self.isochrones]
        if len(levels) != len(set(levels)):
            msg = "one isochrone per minutes level"
            raise ValueError(msg)
        return self

    def area_km2(self, minutes: int) -> float | None:
        for iso in self.isochrones:
            if iso.minutes == minutes:
                return iso.area_km2
        return None

    @computed_field  # type: ignore[prop-decorator]
    @property
    def catchment_ratio(self) -> float | None:
        """15-min area divided by the dry baseline (the ring gauge)."""
        area = self.area_km2(15)
        if area is None or self.dry_area_15_km2 <= 0:
            return None
        return area / self.dry_area_15_km2

    @computed_field  # type: ignore[prop-decorator]
    @property
    def collapse(self) -> bool:
        """True when the 15-min catchment is below 40 % of the dry baseline."""
        ratio = self.catchment_ratio
        return ratio is not None and ratio < REACH_COLLAPSE_RATIO


class RoadCondition(VarunaModel):
    """Feature properties of the ``road-conditions`` provider feed (impassable / degraded segments
    with validity windows) for navigation and transit apps."""

    segment_id: IdStr
    name: str | None = None
    status: RoadStatus
    profile: VehicleProfile = Field(description="Profile the status applies to.")
    valid_from: Timestamp
    valid_to: Timestamp
    depth_p50_cm: float = Field(ge=0)
    p_exceed: Probability = Field(description="P(depth > profile threshold) in the window.")
    run_id: str

    @model_validator(mode="after")
    def _window(self) -> RoadCondition:
        if self.valid_to < self.valid_from:
            msg = "valid_to must not be before valid_from"
            raise ValueError(msg)
        return self


__all__ = [
    "AvoidedSegment",
    "Corridor",
    "Isochrone",
    "ReachMinutes",
    "ReachabilityResponse",
    "ReasonKind",
    "RoadCondition",
    "RoadStatus",
    "RouteConfidence",
    "RouteLabel",
    "RouteReason",
    "RouteRequest",
    "RouteResponse",
    "RouteResult",
]

File: apps/command/lib/api/route.ts
/**
 * Routing and reachability (`POST /v1/route`, `GET /v1/reachability`; CLAUDE.md 7.4, 12).
 *
 * Places are resolved from the city's own layers rather than from coordinates typed into this
 * file: hospitals and fire stations come from the asset register, junctions from the chronic
 * hotspot register, and both carry a `source_url` there (CLAUDE.md 7). A hard-coded lon/lat for
 * "Hindmata" would be a number nobody can check.
 */

import { apiUrl } from "@/lib/api/client";

export interface Place {
  id: string;
  name: string;
  kind: "hospital" | "fire_station" | "hotspot";
  lon: number;
  lat: number;
}

export interface RouteLeg {
  minutes: number;
  distanceM: number;
  maxDepthCm: number;
  depart: string;
  arrive: string;
  safeUntil: string | null;
  path: [number, number][];
  streets: string[];
}

export interface RouteAvoided {
  segmentId: string;
  name: string;
  depthCm: number;
  probability: number;
  at: string;
  /** The street's own geometry, so the map can draw what the detour went around. */
  path: [number, number][];
}

/**
 * One of up to three safe roads the policy spreads traffic across (TECH_SPEC 3.3).
 *
 * `share` is a policy, not a measured traffic count, and the screen must say so. An API that
 * predates route spreading returns none of these, and the screen then shows one answer.
 */
export interface RouteCorridor {
  id: string;
  label: string;
  route: RouteLeg | null;
  share: number;
  assigned: boolean;
  capacityScore: number;
  maxProbability: number;
}

export type RouteReasonKind = "avoided" | "design" | "timing" | "closure";

/**
 * Why the route went this way, as a record rather than a sentence: `lib/explain.ts` words it.
 *
 * Every field beyond `kind` is optional on the wire, because a reason whose number the run could
 * not supply is dropped rather than softened (UI_SPEC 4) - and because an older API sends none.
 */
export interface RouteReason {
  kind: RouteReasonKind;
  segmentId: string;
  name: string;
  depthCm?: number;
  thresholdCm?: number;
  at?: string;
  probability?: number;
  designIntensityMmH?: number;
  forecastPeakMmH?: number;
  dryUntil?: string;
  dryBelowCm?: number;
  reason?: string;
  user?: string;
  until?: string | null;
}

export interface RoutePlan {
  runId: string;
  profile: string;
  departAt: string;
  naive: RouteLeg | null;
  varuna: RouteLeg | null;
  alternates: RouteLeg[];
  avoided: RouteAvoided[];
  /** Empty when the API does not spread, or when only one safe road exists. */
  corridors: RouteCorridor[];
  /** Empty when the API does not explain, or when no reason kept all of its numbers. */
  reasons: RouteReason[];
  tripId: string | null;
  notes: string[];
  ms: number;
}

export interface ReachabilityBand {
  minutes: number;
  areaKm2: number;
  dryAreaKm2: number;
  nJunctions: number;
  nJunctionsDry: number;
  rings: [number, number][][];
}

export interface ReachabilityResult {
  runId: string;
  facility: Place;
  validTs: string;
  profile: string;
  collapsed: boolean;
  shareOfDry: number;
  bands: ReachabilityBand[];
  ms: number;
}

function leg(raw: Record<string, unknown> | null | undefined): RouteLeg | null {
  if (!raw) return null;
  return {
    minutes: Number(raw.minutes ?? 0),
    distanceM: Number(raw.distance_m ?? 0),
    maxDepthCm: Number(raw.max_depth_cm ?? 0),
    depart: String(raw.depart ?? ""),
    arrive: String(raw.arrive ?? ""),
    safeUntil: (raw.safe_until as string | null) ?? null,
    path: (raw.path as [number, number][]) ?? [],
    streets: (raw.streets as string[]) ?? [],
  };
}

/** Every place the trip pickers offer, hospitals first then chronic junctions. */
export async function loadPlaces(city = "mumbai", signal?: AbortSignal): Promise<Place[]> {
  const [facilities, hotspots] = await Promise.all([
    fetch(apiUrl(`/v1/route/facilities?city=${city}`), { signal })
      .then((r) => (r.ok ? r.json() : { facilities: [] }))
      .catch(() => ({ facilities: [] })),
    fetch(apiUrl(`/v1/city/${city}/layers/hotspots`), { signal })
      .then((r) => (r.ok ? r.json() : { features: [] }))
      .catch(() => ({ features: [] })),
  ]);

  const out: Place[] = [];
  for (const f of (facilities.facilities ?? []) as Record<string, unknown>[]) {
    out.push({
      id: String(f.asset_id ?? ""),
      name: String(f.name ?? ""),
      kind: f.kind === "fire_station" ? "fire_station" : "hospital",
      lon: Number(f.lon),
      lat: Number(f.lat),
    });
  }
  for (const feature of (hotspots.features ?? []) as Record<string, never>[]) {
    const props = (feature.properties ?? {}) as Record<string, unknown>;
    const geometry = feature.geometry as { type?: string; coordinates?: number[] } | undefined;
    if (geometry?.type !== "Point" || !geometry.coordinates) continue;
    out.push({
      id: String(props.hotspot_id ?? props.id ?? props.name ?? ""),
      name: String(props.name ?? "Unnamed junction"),
      kind: "hotspot",
      lon: Number(geometry.coordinates[0]),
      lat: Number(geometry.coordinates[1]),
    });
  }
  return out;
}

/**
 * The API's key for a vehicle profile.
 *
 * The console's picker spells two profiles with a hyphen ("two-wheeler", "fire-tender") and
 * `varuna_route.profiles.PROFILES` keys them with an underscore, so sending the picker's value
 * as it stands got a 422 "No vehicle profile 'two-wheeler'" for both. Every other key is
 * already the same on both sides and passes through unchanged.
 */
export function apiProfile(profile: string): string {
  return profile.replace(/-/g, "_");
}

export interface RouteQuery {
  origin: Place;
  destination: Place;
  departAt: string;
  profile: string;
  riskTolerance: number;
  runId?: string;
  /** Ask for up to three corridors. Default on the API side is true. */
  spread?: boolean;
  /** The caller's own id, so repeated requests for one trip keep the same corridor. */
  tripId?: string;
  /** Ask for structured reasons. Default on the API side is true. */
  explain?: boolean;
}

/** Plan a trip. Throws with the API's own message when it refuses one. */
export async function planRoute(query: RouteQuery, signal?: AbortSignal): Promise<RoutePlan> {
  const response = await fetch(apiUrl("/v1/route"), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    signal,
    body: JSON.stringify({
      origin: [query.origin.lon, query.origin.lat],
      destination: [query.destination.lon, query.destination.lat],
      depart_at: query.departAt,
      profile: query.profile,
      risk_tolerance: query.riskTolerance,
      run_id: query.runId,
      spread: query.spread,
      trip_id: query.tripId,
      explain: query.explain,
    }),
  });
  const body = (await response.json()) as Record<string, unknown>;
  if (!response.ok) {
    const envelope = body.error as { message?: string } | undefined;
    throw new Error(envelope?.message ?? `Route failed: HTTP ${response.status}`);
  }
  return {
    runId: String(body.run_id ?? ""),
    profile: String(body.profile ?? ""),
    departAt: String(body.depart_at ?? ""),
    naive: leg(body.naive as Record<string, unknown> | null),
    varuna: leg(body.varuna as Record<string, unknown> | null),
    alternates: ((body.alternates as Record<string, unknown>[]) ?? [])
      .map(leg)
      .filter((r): r is RouteLeg => r !== null),
    avoided: ((body.avoided as Record<string, unknown>[]) ?? []).map((a) => ({
      segmentId: String(a.segment_id ?? ""),
      name: String(a.name ?? ""),
      depthCm: Number(a.depth_cm ?? 0),
      probability: Number(a.probability ?? 0),
      at: String(a.at ?? ""),
      path: (a.path as [number, number][]) ?? [],
    })),
    corridors: ((body.corridors as Record<string, unknown>[]) ?? []).map((c) => ({
      id: String(c.id ?? ""),
      label: String(c.label ?? ""),
      route: leg(c.route as Record<string, unknown> | null),
      share: Number(c.share ?? 0),
      assigned: Boolean(c.assigned),
      capacityScore: Number(c.capacity_score ?? 0),
      maxProbability: Number(c.max_probability ?? 0),
    })),
    reasons: ((body.reasons as Record<string, unknown>[]) ?? [])
      .map(reason)
      .filter((r): r is RouteReason => r !== null),
    tripId: body.trip_id == null ? null : String(body.trip_id),
    notes: (body.notes as string[]) ?? [],
    ms: Number(body.ms ?? 0),
  };
}

const REASON_KINDS: ReadonlySet<string> = new Set(["avoided", "design", "timing", "closure"]);

/**
 * One reason record, or `null` when it is not one of the four kinds `lib/explain.ts` can word.
 *
 * Numbers are carried through only when the wire actually holds them: `undefined` here is what
 * makes the sentence builder drop a reason rather than print "undefined cm".
 */
function reason(raw: Record<string, unknown>): RouteReason | null {
  const kind = String(raw.kind ?? "");
  if (!REASON_KINDS.has(kind)) return null;
  const number = (key: string): number | undefined => {
    const value = raw[key];
    if (value == null) return undefined;
    const n = Number(value);
    return Number.isFinite(n) ? n : undefined;
  };
  const text = (key: string): string | undefined => {
    const value = raw[key];
    return value == null || value === "" ? undefined : String(value);
  };
  return {
    kind: kind as RouteReasonKind,
    segmentId: String(raw.segment_id ?? ""),
    name: String(raw.name ?? ""),
    depthCm: number("depth_cm"),
    thresholdCm: number("threshold_cm"),
    at: text("at"),
    probability: number("probability"),
    designIntensityMmH: number("design_intensity_mm_h"),
    forecastPeakMmH: number("forecast_peak_mm_h"),
    dryUntil: text("dry_until"),
    dryBelowCm: number("dry_below_cm"),
    reason: text("reason"),
    user: text("user"),
    until: raw.until == null ? null : String(raw.until),
  };
}

/** One facility's catchment at one instant. */
export async function loadReachability(
  facility: string,
  at: string,
  profile: string,
  signal?: AbortSignal,
  runId?: string,
): Promise<ReachabilityResult> {
  const params = new URLSearchParams({ facility, t: at, profile });
  if (runId) params.set("run_id", runId);
  const response = await fetch(apiUrl(`/v1/reachability?${params}`), { signal });
  const body = (await response.json()) as Record<string, unknown>;
  if (!response.ok) {
    const envelope = body.error as { message?: string } | undefined;
    throw new Error(envelope?.message ?? `Reachability failed: HTTP ${response.status}`);
  }
  const f = (body.facility ?? {}) as Record<string, unknown>;
  return {
    runId: String(body.run_id ?? ""),
    facility: {
      id: String(f.asset_id ?? ""),
      name: String(f.name ?? ""),
      kind: f.kind === "fire_station" ? "fire_station" : "hospital",
      lon: Number(f.lon),
      lat: Number(f.lat),
    },
    validTs: String(body.valid_ts ?? ""),
    profile: String(body.profile ?? ""),
    collapsed: Boolean(body.collapsed),
    shareOfDry: Number(body.share_of_dry ?? 1),
    bands: ((body.bands as Record<string, unknown>[]) ?? []).map((b) => ({
      minutes: Number(b.minutes ?? 0),
      areaKm2: Number(b.area_km2 ?? 0),
      dryAreaKm2: Number(b.dry_area_km2 ?? 0),
      nJunctions: Number(b.n_junctions ?? 0),
      nJunctionsDry: Number(b.n_junctions_dry ?? 0),
      rings: (b.rings as [number, number][][]) ?? [],
    })),
    ms: Number(body.ms ?? 0),
  };
}
